### PR TITLE
add a variational autoencoder example

### PR DESCRIPTION
**Pull request by CarloLucibello**
I tested it on both CPU and GPU. Julia 0.6 only. 

Plots could be done a little better...

### CODE REVIEW

**Comment by ilkerkesen** [comment]
- It would be good if we'd have handled the dependencies in somewhere.
- Need to replace MLDatasets package. We've built-in interfaces to load MNIST and Fashion MNIST already.
- Since this is a generative model, it'd be good to show some image generations. I mean, it'd be good to have an additional Jupyter Notebook which demonstrates a sample generation by using a trained model.
- Could we make it with ArgParse module just like other examples? It makes easier to run the scripts by using both command line and Julia REPL.

**Comment by ilkerkesen** [comment]
@CarloLucibello if you don't want to solve the above issues, I could make some of them, too.

**Comment by CarloLucibello** [comment]
I can address the issues when I find some time, except for the notebook probably

**Comment by ilkerkesen** [comment]
@CarloLucibello I can do the notebook, please just make sure [this](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) option is enabled.

**Comment by CarloLucibello** [comment]
should be enabled

**Comment by CarloLucibello** [comment]
@ilkerkesen I updated the code, it's all yours now :) Didn't try on GPU this revised version, it should work but please test it because I can't right now. bye

**Comment by ilkerkesen** [comment]
@CarloLucibello OK, I will test it. Thanks a lot.

**Comment by ilkerkesen** [comment]
I've just added a simple demo for this example, it's ready to be merged.

File: src/data.jl [hunk 1]
```diff
@@ -47,6 +47,10 @@ function Base.length(m::MB)
     m.partial ? ceil(Int,n) : floor(Int,n)
 end
 
+function Base.rand(m::MB)
+    i = rand(0:(m.length-m.batchsize))
+    return next(m, i)[1]
+end
```
Comment (denizyuret) [inline src/data.jl]: Why do we need a rand(m::MB) ?  minibatches already have a shuffle option.

**Comment by denizyuret** [comment]
What is Base.rand(m::MB) for?

File: src/data.jl (hunk 1 above)
Comment (ilkerkesen) [reply src/data.jl]: I think it's been used in [here](https://github.com/CarloLucibello/Knet.jl/blob/6182c4f1c6b77ab84613c916cfa6023ff631653f/examples/variational-autoencoder/vae_mnist.jl#L97) to sample a random batch. Nonetheless, the following is equivalent,

```julia
x, _ = first(minibatch(xtrn, ytrn, batchsize, shuffle=true))
```

**Comment by denizyuret** [comment]
OK, the difference is (1) minibatch with shuffle=true shuffles at instance
level, and (2) the rand() solution does not do a reshuffle and is
consistent with rand(::Array). On the other hand rand(::MB) is written such
that it may return a minibatch that a regular for iteration would not
return (because it increases its index in multiples of batchsize).  I will
keep it for now.

**Pull request merged**